Add sample to forEachTween

filiph opened the pull request:
The current documentation for `AnimatedWidgetBaseState.forEachTween` is exhaustive but a little hard to grok. An example will be useful.

## Description

*This adds a short sample to the method.

## Related Issues

* Sorry, I did not create an issue for this. Feel free to push back.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

filiph committed:
Add sample to forEachTween

The current documentation for `AnimatedWidgetBaseState.forEachTween` is exhaustive but a little hard to grok. An example will be useful.

filiph committed:
Solve analysis issue

filiph committed:
Fix the class definition

filiph committed:
Define full class of the ImplicitlyAnimatedWidget

filiph committed:
Remove blank lines from preamble

filiph committed:
Fix preamble

filiph committed:
Remove whitespace

filiph commented:
(Btw, the ten thousand little commits is a testament to the fact that you never want to modify code in the github online editor, however small you think your CL is. Thank god for squash.)

matthew-carroll commented:
@filiph Did you just need a random reviewer, or was there something specific in here that you thought was relevant to me?

filiph commented:
GitHub recommended you in the UI, so I just clicked your name. I though the recommendation is based on the latest person to touch the modified lines or something.

But yeah, I mostly need a random reviewer.

matthew-carroll commented:
Heh, interesting.  Do you know CI is failing for this PR?

matthew-carroll commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -15,6 +15,14 @@ import 'text.dart';
 import 'ticker_provider.dart';
 import 'transitions.dart';
 
+// Examples can assume:
```
Can you include some context around this comment?  It is probably obvious to people who work in the framework docs a lot, but for others this feels random and strange.

Perhaps you can explain that some docs in this class have source code examples, and those source code examples assume the existence of `MyWidget`.

Also, is this the correct place such a comment?  Do we do this elsewhere where we add a double slash `//` multiline comment above the class comment?

filiph replied:
This is a magic invocation defined by https://github.com/flutter/flutter/blob/master/dev/bots/analyze-sample-code.dart. [Other such preambles](https://github.com/flutter/flutter/search?q=%22Examples+can+assume%22&unscoped_q=%22Examples+can+assume%22) are not documented, so I didn't feel I should document this one.

matthew-carroll commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -15,6 +15,14 @@ import 'text.dart';
+// Examples can assume:
+// class MyWidget extends ImplicitlyAnimatedWidget {
+//   MyWidget() : super(duration: const Duration(seconds: 1));
+//   final Color targetColor = Colors.black;
```
Can you add a blank line between members for clarity?

filiph replied:
Unfortunately, that seems not to be supported by the sample tool. When I had a blank line here, the analyzer step failed with `Error: Unexpected content in sample code preamble.`

matthew-carroll commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -358,6 +366,31 @@ abstract class ImplicitlyAnimatedWidgetState<T extends ImplicitlyAnimatedWidget>
   /// properties. Dependent properties should not be updated within
   /// [forEachTween].
+  ///
+  /// {@tool sample}
```
I think we have a policy about always placing a `#Sample Code` header above sample code, right?

matthew-carroll replied:
Also, it would be good to state under the header what this example is intended to show.

filiph replied:
I assumed `# Sample code` is just for classes and constructors but that's stupid. Will do.

matthew-carroll commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -358,6 +366,31 @@ abstract class ImplicitlyAnimatedWidgetState<T extends ImplicitlyAnimatedWidget>
+  ///
+  ///   @override
+  ///   void forEachTween(TweenVisitor<dynamic> visitor) {
+  ///     _colorTween = visitor(
```
Can you offer a brief comment in the sample code that illuminates the purpose of each line here?  I find the visitor system within implicit animations to appear very perplexing and I always seem to forget what it is that I'm trying to accomplish when I write `forEachTween()`.

filiph replied:
Will do.

matthew-carroll reviewed (approved):
Approved, pending comments.

filiph committed:
Address code review feedback

filiph committed:
Remove Sample section

The sample analyzer tool fails with `Error: Found deprecated '## Sample code' section: use {@tool sample}...{@end-tool} instead.` Reverting to previous state.

filiph commented:
Ok, I addressed what I could. Apparently, the `### Sample` section is deprecated, so I reverted back to just using `{@tool sample}`.

```
Error: Found deprecated '## Sample code' section: use {@tool sample}...{@end-tool} instead.
```

matthew-carroll commented:
Interesting.  Ok, LGTM.

goderbauer commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
````diff
@@ -358,6 +366,40 @@ abstract class ImplicitlyAnimatedWidgetState<T extends ImplicitlyAnimatedWidget>
   /// [forEachTween].
+  ///
+  /// {@tool sample}
+  /// ```dart
````
It's good practice to give a one or two sentence introduction within the `{@tool sample}` section that briefly explains in prose what the example code is accomplishing.

filiph replied:
Ack. Done.

filiph committed:
Re-add sample description
